File: notifier.rb
#!/usr/bin/env ruby

require 'bunny'
require 'json'
require 'yaml'
require 'slack/incoming/webhooks'

config_file = ARGV[0] || File.dirname(__FILE__) + '/config.yml'
config = YAML.load_file(config_file)

def rabbitmq_init(config)
  host = config['rabbitmq_host']
  port = config['rabbitmq_port'] || 5672
  user = config['rabbitmq_user'] || 'guest'
  pass = config['rabbitmq_password'] || 'guest'

  Bunny.new(hostname: host, port: port, user: user, pass: pass)
end

webhook_url = config["slack_webhook_url"]
slack = Slack::Incoming::Webhooks.new(webhook_url)

connection = rabbitmq_init(config)
connection.start

ch = connection.create_channel
q = ch.queue("live-migration_notifier", exclusive: true)
x = ch.topic("nova")
q.bind(x, routing_key: "compute.*")

begin
  q.subscribe(manual_ack: false, block: true) do |delivery_info, properties, msg|
    payload = JSON.parse(JSON.parse(msg)["oslo.message"])
    method = payload["method"]

    case method
    when "pre_live_migration"
      instance = payload["args"]["instance"]["nova_object.data"]

      src = instance["host"]
      dst = delivery_info.routing_key.gsub(/^compute\./, "")

      hostname = instance["display_name"]
      user = payload["_context_user_name"]

      attachment = [{
        text: "live-migrationが開始されました",
        fallback: "start: #{hostname} (#{src} -> #{dst})",
        color: "warning",
        fields: [
          {
            title: "instance",
            value: hostname,
            short: true,
          },
          {
            title: "user",
            value: user,
            short: true,
          },
          {
            title: "src",
            value: src,
            short: true
          },
          {
            title: "dst",
            value: dst,
            short: true
          },
        ]
      }]
    when "post_live_migration_at_destination"
      instance = payload["args"]["instance"]["nova_object.data"]

      src = instance["host"]
      dst = delivery_info.routing_key.gsub(/^compute\./, "")

      hostname = instance["display_name"]
      user = payload["_context_user_name"]

      attachment = [{
        text: "live-migrationが完了しました",
        fallback: "end: #{hostname} (#{dst})",
        color: "good",
        fields: [
          {
            title: "instance",
            value: hostname,
            short: true,
          },
          {
            title: "user",
            value: user,
            short: true,
          },
          {
            title: "src",
            value: src,
            short: true
          },
          {
            title: "dst",
            value: dst,
            short: true
          },
        ]
      }]
    when "rollback_live_migration_at_destination"
      instance = payload["args"]["instance"]["nova_object.data"]

      src = instance["host"]
      dst = delivery_info.routing_key.gsub(/^compute\./, "")

      hostname = instance["display_name"]
      instance_uuid = instance["uuid"]
      user = payload["_context_user_name"]
      req_id = payload["_context_request_id"]

      attachment = [{
        text: "live-migrationが失敗しました",
        fallback: "error: #{hostname} (#{dst})",
        color: "danger",
        fields: [
          {
            title: "instance",
            value: hostname,
            short: true,
          },
          {
            title: "user",
            value: user,
            short: true,
          },
          {
            title: "src",
            value: src,
            short: true
          },
          {
            title: "dst",
            value: dst,
            short: true
          },
          {
            title: "request-id",
            value: req_id,
            short: false,
          },
          {
            title: "uuid",
            value: instance_uuid,
            short: false,
          }
        ]
      }]
    end
    slack.post "", attachments: attachment if attachment
  end
rescue Interrupt
  connection.close
end

File: notifier/notifier.go
package notifier

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"

	"github.com/streadway/amqp"
)

type Notifier struct {
	uri         string
	webhook_url string
}

type OsloMessage struct {
	Message string `json:"oslo.message"`
	Version string `json:"oslo.version"`
}

type Payload struct {
	Args             PayloadArgs `json:"args"`
	Method           string      `json:"method"`
	Version          string      `json:"version"`
	ContextUserName  string      `json:"_context_user_name"`
	ContextRequestId string      `json:"_context_request_id"`
}

type PayloadArgs struct {
	Dest     string     `json:"dest"`
	Instance NovaObject `json:"instance"`
}

type NovaObject struct {
	Changes   []string               `json:"nova_object.changes"`
	Data      map[string]interface{} `json:"nova_object.data"`
	Name      string                 `json:"nova_object.name"`
	Namespace string                 `json:"nova_object.namespace"`
	Version   string                 `json:"nova_object.version"`
}

type LiveMigrationData struct {
	Message   string
	Color     string
	Src       string
	Dest      string
	Hostname  string
	User      string
	RequestId string
	UUID      string
}

func Init(uri, webhook_url string) *Notifier {
	return &Notifier{
		uri:         uri,
		webhook_url: webhook_url,
	}
}

func (n *Notifier) Start() error {
	conn, err := amqp.Dial(n.uri)
	if err != nil {
		return err
	}
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil {
		return err
	}
	defer ch.Close()

	queue, err := ch.QueueDeclare("live-migration-notifier", false, false, true, false, nil)
	if err != nil {
		return err
	}

	err = ch.QueueBind(queue.Name, "compute.*", "nova", false, nil)
	if err != nil {
		return err
	}

	msgs, err := ch.Consume(queue.Name, "", true, false, false, false, nil)
	forever := make(chan bool)

	go func() {
		for d := range msgs {
			payload, err := extractPayload(d.Body)
			if err != nil {
				fmt.Fprintf(os.Stderr, "Error: %s\n", err)
				continue
			}

			dest := strings.TrimPrefix(d.RoutingKey, "compute.")
			data := extractMessage(dest, payload)
			if data != nil {
				postSlack(data, n.webhook_url)
			}
		}
	}()

	<-forever
	return nil
}

func extractPayload(data []byte) (*Payload, error) {
	var omsg OsloMessage
	err := json.Unmarshal(data, &omsg)
	if err != nil {
		return nil, err
	}

	var payload Payload
	err = json.Unmarshal([]byte(omsg.Message), &payload)
	if err != nil {
		return nil, err
	}

	return &payload, nil
}

func extractMessage(dest string, payload *Payload) *LiveMigrationData {
	var data LiveMigrationData

	switch payload.Method {
	case "live_migration":
		data.Message = "live-migrationが開始しました"
		data.Color = "warning"

	case "post_live_migration_at_destination":
		data.Message = "live-migrationが完了しました"
		data.Color = "good"

	case "rollback_live_migration_at_destination":
		data.Message = "live-migrationが失敗しました"
		data.Color = "danger"

	default:
		return nil
	}

	data.Src = payload.Args.Instance.Data["host"].(string)
	if payload.Args.Dest != "" {
		data.Dest = payload.Args.Dest
	} else {
		data.Dest = dest
	}
	data.Hostname = payload.Args.Instance.Data["display_name"].(string)
	data.User = payload.ContextUserName
	data.RequestId = payload.ContextRequestId
	data.UUID = payload.Args.Instance.Data["uuid"].(string)

	return &data
}
